### PR TITLE
[Dns-sd] Added Operational Discovery and made browse delegate common

### PR DESCRIPTION
**Pull request by su-shanka**
Added support for operational node browse/discovery. Also, updated existing browse/discover interface in Resolver class to a common interface for all types of node browse and made the delegates common for all browse(commissionable, commissioner and operational)

This is continuation of work in #26718.

### CODE REVIEW

File: src/lib/dnssd/Discovery_ImplPlatform.cpp [hunk 1]
```diff
@@ -743,6 +743,27 @@ CHIP_ERROR DiscoveryImplPlatform::DiscoverCommissioners(DiscoveryFilter filter,
...
 }
 
+CHIP_ERROR DiscoveryImplPlatform::StartDiscovery(DiscoveryType type, DiscoveryFilter filter, DiscoveryContext & context)
+{
+    CHIP_ERROR err = CHIP_ERROR_INVALID_ARGUMENT;
+
+    switch (type)
+    {
+    case DiscoveryType::kCommissionableNode:
+        err = DiscoverCommissionableNodes(filter, context);
```
Comment (andy31415) [inline src/lib/dnssd/Discovery_ImplPlatform.cpp:753]: let's do `return DiscoverCommissionableNodes(filter, context)` and similar below. Make is clearer and uses up less lines:

```cpp
switch (type)
{
    case DiscoveryType::kCommissionableNode:
        return DiscoverCommissionableNodes(filter, context);
    case DiscoveryType::kCommissionerNode:
        return DiscoverCommissioners(filter, context);
    case DiscoveryType::kOperational:
        return CHIP_ERROR_NOT_IMPLEMENTED;
    default:
        return CHIP_ERROR_INVALID_ARGUMENT;
}
```

File: src/lib/dnssd/Types.h [hunk 2]
```diff
@@ -204,29 +204,31 @@ inline constexpr size_t kMaxDeviceNameLen         = 32;
...
+    size_t rotatingIdLen                  = 0;
+    uint32_t deviceType                   = 0;
+    uint16_t longDiscriminator            = 0;
+    uint16_t vendorId                     = 0;
+    uint16_t productId                    = 0;
+    uint16_t pairingHint                  = 0;
+    uint8_t commissioningMode             = 0;
+    uint8_t commissionerPasscode          = 0;
+    uint8_t rotatingId[kMaxRotatingIdLen] = {};
+    bool nodeStatus; // true is ttl > 0 false id ttl = 0
```
Comment (andy31415) [inline src/lib/dnssd/Types.h:220]: This needs a better name. how about `hasZeroTTL` (but then condition is inverted) or something similar?

I should not need to read comments to understand what a value means.

**Review by andy31415 — APPROVED** [review]

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 3]
```diff
@@ -199,10 +182,11 @@ CHIP_ERROR IncrementalResolver::InitializeParsing(mdns::Minimal::SerializedQName
                 return CHIP_ERROR_INVALID_ARGUMENT;
             }
 
-            Platform::CopyString(mSpecificResolutionData.Get<CommissionNodeData>().instanceName, nameCopy.Value());
+            Platform::CopyString(mSpecificResolutionData.Get<DnssdNodeData>().instanceName, nameCopy.Value());
         }
+        mSpecificResolutionData.Get<DnssdNodeData>().nodeStatus = (ttl == 0) ? false : true;
```
[the remark this answers is not in the document]
Comment (su-shanka) [reply src/lib/dnssd/IncrementalResolve.cpp:187]: Sure.

File: src/lib/dnssd/Types.h (hunk 2 above)
Comment (su-shanka) [reply src/lib/dnssd/Types.h:220]: That makes sense, application can decide how to use the info of ttl being 0. Will update it to `hasZeroTTL`.

File: src/lib/dnssd/Discovery_ImplPlatform.cpp (hunk 1 above)
Comment (su-shanka) [reply src/lib/dnssd/Discovery_ImplPlatform.cpp]: updated

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Please break out the "pure rename, no functional changes" bits into a separate PR from the "now we are changing behavior" bits.  That will make it much easier to review both PRs.

**Comment by su-shanka** [comment]
> Please break out the "pure rename, no functional changes" bits into a separate PR from the "now we are changing behavior" bits. That will make it much easier to review both PRs.

Moved the refactor and renaming changes to #32866. Will update this PR once #32866  is closed.

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp [hunk 4]
```diff
@@ -418,6 +418,10 @@ void MinMdnsResolver::AdvancePendingResolverStates()
                 mActiveResolves.CompleteCommissionable(nodeData);
                 nodeData.nodeType = DiscoveryType::kCommissionableNode;
                 break;
+            case IncrementalResolver::ServiceNameType::kOperational:
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:421]: Is there something tracking adding this to the "platform" dns-sd implementation?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Requesting changes for now until the issues in the underlying PR are resolved.

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 4 above)
Comment (su-shanka) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:421]: Since the platform implementation of operational discovery in #28301 is blocked on some discussion, it is not considered here.

**Review by andy31415 — CHANGES_REQUESTED** [review]
Blocking until https://github.com/project-chip/connectedhomeip/pull/32866/files#r1561854542 is resolved

We need to have resoltion data be "valid data" without a large structure with members that are a union of all possible members.

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 5]
```diff
@@ -201,6 +184,7 @@ CHIP_ERROR IncrementalResolver::InitializeParsing(mdns::Minimal::SerializedQName
 
             Platform::CopyString(mSpecificResolutionData.Get<DnssdNodeData>().instanceName, nameCopy.Value());
         }
+        mSpecificResolutionData.Get<DnssdNodeData>().hasZeroTTL = (ttl == 0) ? false : true;
```
Comment (andy31415) [inline src/lib/dnssd/IncrementalResolve.cpp:187]: `(condition) ? false : true` is an antipattern (and in this case it seems reversed. it should be just `condition`

Comment (su-shanka) [reply src/lib/dnssd/IncrementalResolve.cpp]: updated with proper check

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 6]
```diff
@@ -341,17 +344,36 @@ CHIP_ERROR IncrementalResolver::OnIpAddress(Inet::InterfaceId interface, const I
...
+template <class NodeType>
+CHIP_ERROR IncrementalResolver::Set(DiscoveredNodeData & output)
+{
+    VerifyOrReturnError(mSpecificResolutionData.Is<NodeType>(), CHIP_ERROR_INCORRECT_STATE);
+    output.Set<NodeType>();
+    auto & nodeData                       = output.Get<NodeType>();
+    nodeData                              = mSpecificResolutionData.Get<NodeType>();
+    CommonResolutionData & resolutionData = nodeData;
+    resolutionData                        = mCommonResolutionData;
+    return CHIP_NO_ERROR;
```
Comment (andy31415) [inline src/lib/dnssd/IncrementalResolve.cpp:356]: There are several intermediate variables above. Could you re-visit the logic and see if it can be cleaned up? 

As it stands, this feels a bit too much magic in casting and operator=. Why do we have this complex of code?

Comment (andy31415) [reply src/lib/dnssd/IncrementalResolve.cpp:356]: At a minimum `output.Set<NodeType>(mSpecificResolutionData.Get<NodeType>())` makes some of the assignments more explicity ... however I seem to get lost in what mSpecificResolutionData vs output are supposed to contain and their difference in types.

Comment (andy31415) [reply src/lib/dnssd/IncrementalResolve.cpp:356]: I see this was copied & pasted from previous code, however previous code was also part of the refactor. It used to be:

```cpp
outputData.resolutionData = mCommonResolutionData;
outputData.nodeData       = mSpecificResolutionData.Get<DnssdNodeData>();
```

however now it seems significantly messier. This needs fixing to be readable.

File: src/lib/shell/commands/Dns.cpp [hunk 7]
```diff
@@ -83,26 +83,47 @@ class DnsShellResolverDelegate : public Dnssd::DiscoverNodeDelegate, public Addr
...
-        Dnssd::CommissionNodeData nodeData = discNodeData.Get<Dnssd::CommissionNodeData>();
+        auto nodeData = discNodeData.Get<Dnssd::CommissionNodeData>();
+
         if (!nodeData.IsValid())
         {
             streamer_printf(streamer_get(), "DNS browse failed - not found valid services \r\n");
             return;
         }
 
+        streamer_printf(streamer_get(), "DNS browse succeeded: \r\n");
```
Comment (Damian-Nordic) [inline src/lib/shell/commands/Dns.cpp:121]: Why was this moved?

File: src/lib/shell/commands/Dns.cpp [hunk 8]
```diff
@@ -83,26 +83,47 @@ class DnsShellResolverDelegate : public Dnssd::DiscoverNodeDelegate, public Addr
...
+
+        if (discNodeData.Is<Dnssd::OperationalNodeData>())
         {
-            streamer_printf(streamer_get(), "DNS browse failed - not commission type node \r\n");
+            LogOperationalNodeDiscovered(discNodeData.Get<Dnssd::OperationalNodeData>());
             return;
         }
 
-        Dnssd::CommissionNodeData nodeData = discNodeData.Get<Dnssd::CommissionNodeData>();
+        auto nodeData = discNodeData.Get<Dnssd::CommissionNodeData>();
```
Comment (Damian-Nordic) [inline src/lib/shell/commands/Dns.cpp:113]:
```suggestion
        const auto& nodeData = discNodeData.Get<Dnssd::CommissionNodeData>();
```

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 9]
```diff
@@ -167,6 +168,7 @@ CHIP_ERROR IncrementalResolver::InitializeParsing(mdns::Minimal::SerializedQName
 
     switch (mServiceNameType)
     {
+    /* All known dnssd records for MATTER are filled into DnssdNodeData */
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/IncrementalResolve.cpp:171]: What is DnssdNodeData?

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp [hunk 10]
```diff
@@ -452,18 +452,49 @@ void MinMdnsResolver::AdvancePendingResolverStates()
...
+            ResolvedNodeData nodeResolvedData;
+            CHIP_ERROR err = CHIP_ERROR_UNINITIALIZED;
+
+            if (mActiveResolves.HasBrowseFor(chip::Dnssd::DiscoveryType::kCommissionableNode))
+            {
+                DiscoveredNodeData nodeData;
+                err = resolver->Take(nodeData);
+                if (!nodeData.Is<OperationalNodeData>())
+                {
+                    err = CHIP_ERROR_INVALID_ARGUMENT;
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:464]: Isn't this losing the `err` that `Take` might have returned?

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp [hunk 11]
```diff
@@ -452,18 +452,49 @@ void MinMdnsResolver::AdvancePendingResolverStates()
...
+                {
+                    mDiscoveryContext->OnNodeDiscovered(nodeData);
+                    auto opNodeData                  = nodeData.Get<OperationalNodeData>();
+                    nodeResolvedData.resolutionData  = static_cast<CommonResolutionData>(opNodeData);
+                    nodeResolvedData.operationalData = opNodeData;
+                }
+                else
+                {
+#if CHIP_MINMDNS_HIGH_VERBOSITY
+                    ChipLogError(Discovery, "No delegate to report commissioning node discovery");
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:480]: This log does not look correct; we might have a delegate, but `err` could be a failure.

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp [hunk 12]
```diff
@@ -452,18 +452,49 @@ void MinMdnsResolver::AdvancePendingResolverStates()
...
+                }
+                if (err != CHIP_NO_ERROR)
+                {
+                    ChipLogError(Discovery, "Failed to take discovery result: %" CHIP_ERROR_FORMAT, err.Format());
+                }
+                if (mDiscoveryContext != nullptr && err == CHIP_NO_ERROR)
+                {
+                    mDiscoveryContext->OnNodeDiscovered(nodeData);
+                    auto opNodeData                  = nodeData.Get<OperationalNodeData>();
+                    nodeResolvedData.resolutionData  = static_cast<CommonResolutionData>(opNodeData);
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:474]: Why is this cast needed?

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp [hunk 13]
```diff
@@ -452,18 +452,49 @@ void MinMdnsResolver::AdvancePendingResolverStates()
...
             MATTER_TRACE_SCOPE("Active operational delegate call", "MinMdnsResolver");
-            ResolvedNodeData nodeData;
+            ResolvedNodeData nodeResolvedData;
+            CHIP_ERROR err = CHIP_ERROR_UNINITIALIZED;
+
+            if (mActiveResolves.HasBrowseFor(chip::Dnssd::DiscoveryType::kCommissionableNode))
+            {
+                DiscoveredNodeData nodeData;
+                err = resolver->Take(nodeData);
+                if (!nodeData.Is<OperationalNodeData>())
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:462]: I don't understand this.  We have a browse for kCommissionableNode; why are we suddenly checking for operational here?

File: src/lib/dnssd/Types.h [hunk 14]
```diff
@@ -194,11 +194,15 @@ struct CommonResolutionData
 };
 
 /// Data that is specific to Operational Discovery of nodes
-struct OperationalNodeData
+struct OperationalNodeData : public CommonResolutionData
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Types.h:197]: Hmm, so now `ResolvedNodeData` will basically have two copies of `CommonResolutionData`?  One direct, one via `OperationalNodeData`?

That's really confusing as an API consumer.  Do those copies have to be in sync?  If not, which one is the one to use?  If yes, why do we have both?

Seems like there's another intermediate change needed here which is to make this inheritance change and change `ResolvedNodeData` to either just have an `OperationalNodeData` member, or inherit from `OperationalNodeData` and have no data members, or just be an alias for `OperationalNodeData` (and figure out the logging situation).

File: src/lib/dnssd/Types.h [hunk 15]
```diff
@@ -194,11 +194,15 @@ struct CommonResolutionData
...
 {
     PeerId peerId;
-
-    void Reset() { peerId = PeerId(); }
+    bool hasZeroTTL;
```
[the remark this answers is not in the document]
Comment (su-shanka) [reply src/lib/dnssd/Types.h:200]: The zero TTL in operational indicates that it has been removed from or left the fabric, hence this helps detect when a operational node joins and leaves the respective fabric.

File: src/lib/dnssd/IncrementalResolve.cpp (hunk 9 above)
Comment (su-shanka) [reply src/lib/dnssd/IncrementalResolve.cpp:171]: Seems to have sneaked in during auto merge of rebase. Will correct it.

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 13 above)
Comment (su-shanka) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:462]: It was supposed to be operational browse check, fixed

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 10 above)
Comment (su-shanka) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:464]: Yes, but if there is no OperationalNodeData available then even if `Take` returned success we cannot use it. If the data is of type OperationalNodeData only then the `err` from `Take` has significance. But I guess I should do that type check only is `Take` is successful. I will update the logic a bit.

File: src/lib/dnssd/Types.h (hunk 14 above)
Comment (su-shanka) [reply src/lib/dnssd/Types.h:197]: I was actually wondering if we need to provide resolution data in delegate for Operational Discovery. From my understanding, the idea of Operational Browse is to detect new nodes in fabric, application follow up with a resolution request if resolution data is needed for any particular discovered node. Any thoughts on this?

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 12 above)
Comment (su-shanka) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:474]: this part removed as no longer needed.

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 11 above)
Comment (su-shanka) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:480]: fixed

File: src/lib/dnssd/Types.h (hunk 14 above)
Comment (su-shanka) [reply src/lib/dnssd/Types.h]: As agreed upon, removed inheritance of `CommonResolutionData` from `OperationalNodeData`  as it is not a necessity for Operational Browse.

File: src/lib/shell/commands/Dns.cpp [hunk 16]
```diff
@@ -83,26 +83,47 @@ class DnsShellResolverDelegate : public Dnssd::DiscoverNodeDelegate, public Addr
...
+        streamer_printf(streamer_get(), "DNS browse operational succeeded: \r\n");
+        streamer_printf(streamer_get(), "   Node ID: " ChipLogFormatX64 "-" ChipLogFormatX64 "\r\n",
+                        ChipLogValueX64(nodeData.peerId.GetCompressedFabricId()), ChipLogValueX64(nodeData.peerId.GetNodeId()));
+        streamer_printf(streamer_get(), "   hasZeroTTL: %s\r\n", nodeData.hasZeroTTL ? "true" : "false");
+    }
+
     void OnNodeDiscovered(const Dnssd::DiscoveredNodeData & discNodeData) override
     {
-        if (!discNodeData.Is<Dnssd::CommissionNodeData>())
+        if (!discNodeData.Is<Dnssd::CommissionNodeData>() && !discNodeData.Is<Dnssd::OperationalNodeData>())
```
[the remark this answers is not in the document]
Comment (su-shanka) [reply src/lib/shell/commands/Dns.cpp:101]: removed.

File: src/lib/shell/commands/Dns.cpp (hunk 8 above)
Comment (su-shanka) [reply src/lib/shell/commands/Dns.cpp:113]: done

File: src/lib/shell/commands/Dns.cpp (hunk 7 above)
Comment (su-shanka) [reply src/lib/shell/commands/Dns.cpp:121]: I moved it to ensure to have common code for both node types, which was reverted later on, missed to revert this moved.
Reverted it now.

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 10 above)
Comment (su-shanka) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp]: this logic was no longer needed with removal of inheritance. removed it.

File: src/lib/dnssd/IncrementalResolve.cpp (hunk 6 above)
Comment (su-shanka) [reply src/lib/dnssd/IncrementalResolve.cpp]: changed this a bit.

File: src/lib/dnssd/IncrementalResolve.cpp (hunk 9 above)
Comment (su-shanka) [reply src/lib/dnssd/IncrementalResolve.cpp]: removed the comment.

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 17]
```diff
@@ -343,14 +345,15 @@ CHIP_ERROR IncrementalResolver::OnIpAddress(Inet::InterfaceId interface, const I
...
     IPAddressSorter::Sort(mCommonResolutionData.ipAddress, mCommonResolutionData.numIPs, mCommonResolutionData.interfaceId);
 
-    outputData.Set<CommissionNodeData>();
-    CommissionNodeData & nodeData         = outputData.Get<CommissionNodeData>();
-    nodeData                              = mSpecificResolutionData.Get<CommissionNodeData>();
-    CommonResolutionData & resolutionData = nodeData;
+    // Set nodeData with data specific to CommissionNodeData
+    outputData.Set<CommissionNodeData>(mSpecificResolutionData.Get<CommissionNodeData>());
+
+    // set the common resolution data to nodeData
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/IncrementalResolve.cpp:355]: I don't understand this comment.  What is `nodeData` here?

In general, this "assign to the same struct twice" needs some clear documentation about the assumptions and ordering here.  

As a followup, the fact that we are keeping `mCommonResolutionData` separately from `mSpecificResolutionData.Get<CommissionNodeData>()` which _also_ holds the common resolution data is pretty confusing and should be fixed.  But not in this PR.

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp [hunk 18]
```diff
@@ -452,18 +453,46 @@ void MinMdnsResolver::AdvancePendingResolverStates()
...
+                    opNodeData.hasZeroTTL = nodeResolvedData.operationalData.hasZeroTTL;
+                    nodeData.Set<OperationalNodeBrowseData>(opNodeData);
+                }
+                if (err != CHIP_NO_ERROR)
+                {
+                    ChipLogError(Discovery, "Failed to take discovery result: %" CHIP_ERROR_FORMAT, err.Format());
+                }
+                else if (mDiscoveryContext != nullptr)
+                {
+                    mDiscoveryContext->OnNodeDiscovered(nodeData);
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:477]: I do have to ask: why are we not just setting up `DiscoveredNodeData` in this branch?  It's not used anywhere else...

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp [hunk 19]
```diff
@@ -452,18 +453,46 @@ void MinMdnsResolver::AdvancePendingResolverStates()
         else if (resolver->IsActiveOperationalParse())
         {
             MATTER_TRACE_SCOPE("Active operational delegate call", "MinMdnsResolver");
-            ResolvedNodeData nodeData;
+            ResolvedNodeData nodeResolvedData;
+            CHIP_ERROR err = CHIP_ERROR_UNINITIALIZED;
+
+            err = resolver->Take(nodeResolvedData);
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:459]:
```suggestion

            CHIP_ERROR err = resolver->Take(nodeResolvedData);
```

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp [hunk 20]
```diff
@@ -452,18 +453,46 @@ void MinMdnsResolver::AdvancePendingResolverStates()
...
+                    ChipLogError(Discovery, "No delegate to report operational node discovery");
+#endif
+                }
+            }
 
-            CHIP_ERROR err = resolver->Take(nodeData);
             if (err != CHIP_NO_ERROR)
             {
-                ChipLogError(Discovery, "Failed to take discovery result: %" CHIP_ERROR_FORMAT, err.Format());
+                ChipLogError(Discovery, "Failed to take ResolvedNodeData - result: %" CHIP_ERROR_FORMAT, err.Format());
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:489]: So if Take() fails we will log the error twice if `HasBrowseFor(chip::Dnssd::DiscoveryType::kOperational)`?  Seem confusing.

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp [hunk 21]
```diff
@@ -452,18 +453,46 @@ void MinMdnsResolver::AdvancePendingResolverStates()
...
             }
 
-            mActiveResolves.Complete(nodeData.operationalData.peerId);
+            mActiveResolves.Complete(nodeResolvedData.operationalData.peerId);
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:492]: It's really weird to be doing this if err is not success, but it predates your PR.  Please file a followup issue on this use of `nodeResolvedData` when it might contain arbitrary garbage.

File: src/lib/dnssd/Types.h [hunk 22]
```diff
@@ -197,10 +197,22 @@ struct CommonResolutionData
...
     void Reset() { peerId = PeerId(); }
 };
 
+struct OperationalNodeBrowseData : public OperationalNodeData
+{
+    OperationalNodeBrowseData() { Reset(); };
+    void LogDetail() const
+    {
+        ChipLogDetail(Discovery, "Discovered Operational node:\r\n");
+        ChipLogDetail(Discovery, "\tNode ID: " ChipLogFormatX64 "-" ChipLogFormatX64 "\r\n",
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Types.h:210]: This is really not the "Node ID"... Maybe "Node instance:"?

But also, this seems to be re-inventing ChipLogFormatPeerId/ChipLogValuePeerId.

File: src/lib/shell/commands/Dns.cpp [hunk 23]
```diff
@@ -83,15 +83,24 @@ class DnsShellResolverDelegate : public Dnssd::DiscoverNodeDelegate, public Addr
 
     AddressResolve::NodeLookupHandle & Handle() { return mSelfHandle; }
 
+    void LogOperationalNodeDiscovered(const Dnssd::OperationalNodeBrowseData & nodeData)
+    {
+        streamer_printf(streamer_get(), "DNS browse operational succeeded: \r\n");
+        streamer_printf(streamer_get(), "   Node ID: " ChipLogFormatX64 "-" ChipLogFormatX64 "\r\n",
```
Comment (bzbarsky-apple) [inline src/lib/shell/commands/Dns.cpp:89]: See comments above.

File: src/lib/dnssd/IncrementalResolve.cpp (hunk 17 above)
Comment (su-shanka) [reply src/lib/dnssd/IncrementalResolve.cpp:355]: Added more descriptive comment to explain the changes done.

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 18 above)
Comment (su-shanka) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:477]: it was kept out as per previous logic, it no longer need to be branched that way. I have moved it to `err != CHIP_NO_ERROR` branch

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 19 above)
Comment (su-shanka) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:459]: done

File: src/lib/dnssd/Types.h (hunk 22 above)
Comment (su-shanka) [reply src/lib/dnssd/Types.h:210]: renamed and used ChipLogFormatPeerId+ChipLogValuePeerId for logging.

File: src/lib/shell/commands/Dns.cpp (hunk 23 above)
Comment (su-shanka) [reply src/lib/shell/commands/Dns.cpp:89]: Renamed `Node ID` but this going streamer and not chip logging, so cannot use `ChipLogFormatPeerId` and `ChipLogValuePeerId` here

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 20 above)
Comment (su-shanka) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:485]: thought it would be good idea to have separate error logs for browse/discovery and resolution.

File: src/lib/dnssd/IncrementalResolve.cpp (hunk 17 above)
Comment (bzbarsky-apple) [reply src/lib/dnssd/IncrementalResolve.cpp]: Much better, thank you!

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp [hunk 24]
```diff
@@ -452,18 +453,42 @@ void MinMdnsResolver::AdvancePendingResolverStates()
...
+                    ChipLogError(Discovery, "Failed to take discovery result: %" CHIP_ERROR_FORMAT, err.Format());
+                }
+                else if (mDiscoveryContext != nullptr)
+                {
+                    DiscoveredNodeData nodeData;
+                    OperationalNodeBrowseData opNodeData;
+
+                    opNodeData.peerId     = nodeResolvedData.operationalData.peerId;
+                    opNodeData.hasZeroTTL = nodeResolvedData.operationalData.hasZeroTTL;
+                    nodeData.Set<OperationalNodeBrowseData>(opNodeData);
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:472]: I wish we could just Set<>() with no args and then fill it in in-place instead of copying the struct, but if we can't this is fine of course.

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 20 above)
Comment (bzbarsky-apple) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:485]: I think my point is that we will log both `Failed to take discovery result` and `Failed to take ResolvedNodeData` when `DiscoveryType::kOperational`.

Maybe this second `if (err != CHIP_NO_ERROR)` should just be `else if` so it does not get hit when `DiscoveryType::kOperational`?

File: src/lib/shell/commands/Dns.cpp (hunk 23 above)
Comment (bzbarsky-apple) [reply src/lib/shell/commands/Dns.cpp:89]: This seems to still be using "Node ID".

But also, I don't understand the "needs to go into streamer" distinction: `ChipLogFormatPeerId` is a macro that expands out exactly to `ChipLogFormatX64 "-" ChipLogFormatX64`, no?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp [hunk 25]
```diff
@@ -363,14 +363,15 @@ void MinMdnsResolver::AdvancePendingResolverStates()
     {
         if (!resolver->IsActive())
         {
+            ChipLogError(Discovery, "resolver inactive, continue to next");
```
Comment (Damian-Nordic) [inline src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:366]: It's a normal situation in the case of parallel queries, not an error, right?

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 20 above)
Comment (Damian-Nordic) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:485]: I suggest having a single log just after `Take` is executed. Also, after rebasing to https://github.com/project-chip/connectedhomeip/commit/881676de173e779b9471eaae944c33e5780b038c, the code will continue to the next record in such a case, which it should do.

File: src/lib/dnssd/Types.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (Damian-Nordic) [reply src/lib/dnssd/Types.h:200]: OK, my point is that receiving records with TTL 0 makes sense for continuous queries, while for one-shot queries, which the common Matter's DNS-SD layer uses (except for Darwin that implements a custom interface for continuous queries - see their custom `ChipDnssdBrowse` interface), you rather not receive removed records at all than receive those with TTL 0.

Thus, I'm curious in which configuration do you test that.

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 25 above)
Comment (su-shanka) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:366]: removed.

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 20 above)
Comment (su-shanka) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:485]: Moved log just after Take() as suggested.

File: src/lib/shell/commands/Dns.cpp (hunk 23 above)
Comment (su-shanka) [reply src/lib/shell/commands/Dns.cpp:89]: My bad, I got confused. I have updated.

File: src/lib/dnssd/Resolver_ImplMinimalMdns.cpp (hunk 24 above)
Comment (su-shanka) [reply src/lib/dnssd/Resolver_ImplMinimalMdns.cpp:472]: Andy had a comment to use Set<> with args in another place, hence followed the same here.

File: src/lib/dnssd/Types.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (su-shanka) [reply src/lib/dnssd/Types.h:200]: I didn't realize this. I was expecting continuous queries from matter's dnssd. Now that I tested for a longer duration I see that we stop querying after certain retries. However, this TTL 0 info is added for continuous queries.

[the remark this answers is not in the document]
Comment (Damian-Nordic) [reply src/lib/dnssd/Types.h:200]: Resolving for now.

**Review by Damian-Nordic — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]
